src/components/utils/src/threads/thread_win.cc

```diff
@@ -74,27 +73,50 @@ void* Thread::threadFunc(void* arg) {
   DCHECK(thread);
 
   thread->state_lock_.Acquire();
-  thread->thread_created_ = true;
+  thread->thread_state_ = kThreadStateIdle;
   thread->state_cond_.Broadcast();
 
-  while (!thread->finalized_) {
+  // We have special variable for controlling iterations/exiting thread
+  // in order to separate decision logic (continue iterations or exit?)
+  // from controlling while cycle
+  bool continueIterations = true;
+
+  while (continueIterations) {
     SDL_DEBUG("Thread #" << GetCurrentThreadId() << " iteration");
-    while (!thread->do_iteration_) {
+    while (kThreadCommandNone == thread->thread_command_) {
       thread->run_cond_.Wait(thread->state_lock_);
     }
     SDL_DEBUG("Thread #" << GetCurrentThreadId() << " execute. "
                          << "stopped_ = " << thread->stopped_
                          << "; finalized_ = " << thread->finalized_);
-    if (!thread->stopped_ && !thread->finalized_) {
-      thread->isThreadRunning_ = true;
-
-      thread->state_lock_.Release();
-      thread->delegate_->threadMain();
-      thread->state_lock_.Acquire();
 
-      thread->isThreadRunning_ = false;
+    switch (thread->thread_command_) {
+      case kThreadCommandRun:
+        thread->thread_state_ = kThreadStateRunning;
+        pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
+        pthread_testcancel();
+        thread->state_lock_.Release();
+        thread->delegate_->threadMain();
+        thread->state_lock_.Acquire();
+        pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL);
+        thread->thread_state_ = kThreadStateIdle;
+        break;
+
+      case kThreadCommandFinalize:
+        continueIterations = false;
+        break;
+
+      default:
+        // check for not expected command:
+        if ((kThreadCommandNone > thread->thread_command_) &&
```

Comment:
@LuxoftAKutsan, there are three *.cc implementations for thread class and there are other parts duplicated also. I've added it as item in APPLINK-26975.